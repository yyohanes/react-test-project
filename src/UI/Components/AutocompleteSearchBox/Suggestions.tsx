import React, { useEffect, useRef } from 'react'
import classNames from 'classnames'

import { AvailableStyle } from 'app/UI/Theme'
import ListGroup, { ListGroupItem } from '../ListGroup'
import './styles.module.scss'

type Props<OT, O = OT[]> = {
  focusIndex: number
  options: O
  renderOption: (option: OT, index: number, focusIndex: number) => React.ReactNode
  renderFooter?: (options: O) => React.ReactNode
  focusStyle?: AvailableStyle
  onNearBottom?: () => void,
  onSelect?: (option: OT) => void,
}

export default <OT extends unknown>(props: Props<OT>) => {
  const {
    renderOption,
    options,
    focusStyle,
    renderFooter,
    focusIndex,
    onNearBottom,
    onSelect,
  } = props
  const resultsWrapper = useRef<HTMLDivElement>(null)
  const activeItem = useRef<HTMLDivElement>(null)

  const handleOnClick = (option: OT) => {
    if (onSelect) {
      onSelect(option)
    }
  }

  const handleOnResultsScroll = () => {
    if (resultsWrapper.current && onNearBottom) {
      const distanceToBottom = resultsWrapper.current.scrollHeight - (resultsWrapper.current.scrollTop + resultsWrapper.current.offsetHeight)
      if (distanceToBottom <= 120) {
        onNearBottom()
      }
    }
  }

  useEffect(() => {
    if (activeItem.current && resultsWrapper.current) {
      if (activeItem.current.offsetTop + 30 > resultsWrapper.current.offsetHeight + resultsWrapper.current.scrollTop) {
        activeItem.current.scrollIntoView(false)
      } else if (activeItem.current.offsetTop < resultsWrapper.current.scrollTop) {
        activeItem.current.scrollIntoView()
      }
    }
  }, [focusIndex])

  useEffect(() => {
    if (resultsWrapper.current) {
      resultsWrapper.current.addEventListener('scroll', handleOnResultsScroll)
    }

    return function cleanup() {
      if (resultsWrapper.current) {
        resultsWrapper.current.removeEventListener('scroll', handleOnResultsScroll)
      }
    }
  }, [])

  return (
    <div className={classNames(['autocomplete-search-box__suggestions', 'my-2'])}>
      <ListGroup>
        <div ref={resultsWrapper} className="autocomplete-search-box__results">
          {options.map((option, idx) => (
            // We cannot focus to each issue (we want to preserve input focus). So keyboard event will be handled in parent
            // eslint-disable-next-line jsx-a11y/click-events-have-key-events
            <div
              tabIndex={-1}
              role="button"
              onClick={() => handleOnClick(option)}
              key={idx}
              ref={focusIndex === idx ? activeItem : undefined}
            >
              <ListGroupItem colorStyle={focusIndex === idx ? focusStyle : undefined}>
                {renderOption(option, idx, focusIndex)}
              </ListGroupItem>
            </div>
          ))}
        </div>
        {renderFooter && (
          <ListGroupItem key="footer" colorStyle="secondary">
            {renderFooter(options)}
          </ListGroupItem>
        )}
      </ListGroup>
    </div>
  )
}
