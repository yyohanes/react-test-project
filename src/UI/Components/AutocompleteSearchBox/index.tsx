import React, {
  useState,
  useEffect,
  useRef,
  useCallback,
} from 'react'
import classNames from 'classnames'

import { useStyles, AvailableStyle } from 'app/UI/ThemeProvider'
import ListGroup, { ListGroupItem } from '../ListGroup'
import SearchBox from '../SearchBox'
import moduleStyles from './styles.module.scss'

const shortcutWhitelistedKeys = ['control', 'f']

type Props<OT, O = OT[]> = {
  defaultValue?: string | number
  options: O
  renderOption: (option: OT, index: number, focusIndex: number) => React.ReactNode
  renderFooter?: (options: O) => React.ReactNode
  focusStyle?: AvailableStyle
  placeholder?: string
  onChange?: (changedValue: string) => void
}

export default <OT extends unknown>(props: Props<OT>) => {
  const {
    renderOption,
    defaultValue,
    placeholder,
    options,
    focusStyle,
    onChange,
    renderFooter,
  } = props
  const [isInputFocused, setIsInputFocused] = useState<boolean>(false)
  const [pressedKeys, setPressedKeys] = useState<string[]>([])
  const [focusIndex, setFocusIndex] = useState<number>(-1)
  const searchBoxRef = useRef<HTMLInputElement>(null)
  const { styles } = useStyles()

  const isSearchShortcutPressed = shortcutWhitelistedKeys
    .filter(shortcut => pressedKeys.includes(shortcut)).length === 2

  // Keyboard event listeners
  const handleOnInputKeyDown = useCallback(e => {
    if (['arrowup', 'up'].includes(e.key.toLowerCase())) {
      e.preventDefault()

      setFocusIndex(_focusIndex => {
        if (_focusIndex <= 0) {
          return options.length - 1
        }
        return _focusIndex - 1
      })
    } else if (['arrowdown', 'down'].includes(e.key.toLowerCase())) {
      e.preventDefault()

      setFocusIndex(_focusIndex => {
        if (_focusIndex >= options.length - 1) {
          return 0
        }
        return _focusIndex + 1
      })
    }
  }, [options])

  const handleOnSearchShortcut = e => {
    if (!shortcutWhitelistedKeys.includes(e.key.toLowerCase())) {
      return
    }

    if (e.type === 'keydown') {
      setPressedKeys(_pressedKeys => _pressedKeys
        .filter(pressedKey => pressedKey !== e.key.toLowerCase())
        .concat([e.key.toLowerCase()]))
    } else if (e.type === 'keyup') {
      setPressedKeys(_pressedKeys => _pressedKeys
        .filter(pressedKey => pressedKey !== e.key.toLowerCase()))
    }
  }

  const handleOnInputFocus = () => {
    setIsInputFocused(true)
  }

  const handleOnInputBlur = () => {
    setIsInputFocused(false)
  }

  // Reset focus index on options change
  useEffect(() => {
    setFocusIndex(-1)
  }, [options])

  // Attach Up/Down keyboard listener
  useEffect(() => {
    if (searchBoxRef.current) {
      searchBoxRef.current.addEventListener('keydown', handleOnInputKeyDown)
    }

    return function cleanup() {
      if (searchBoxRef.current) {
        searchBoxRef.current.removeEventListener('keydown', handleOnInputKeyDown)
      }
    }
  }, [handleOnInputKeyDown])

  // Check if pressed keys contain ctrl + f
  useEffect(() => {
    if (searchBoxRef.current && isSearchShortcutPressed) {
      searchBoxRef.current.focus()
    }
  }, [pressedKeys])

  // Attach ctrl+f listener
  useEffect(() => {
    if (searchBoxRef.current) {
      searchBoxRef.current.addEventListener('focus', handleOnInputFocus)
      searchBoxRef.current.addEventListener('blur', handleOnInputBlur)
    }

    document.addEventListener('keyup', handleOnSearchShortcut)
    document.addEventListener('keydown', handleOnSearchShortcut)

    return function cleanup() {
      if (searchBoxRef.current) {
        searchBoxRef.current.removeEventListener('focus', handleOnInputFocus)
        searchBoxRef.current.removeEventListener('blur', handleOnInputBlur)
      }

      document.removeEventListener('keyup', handleOnSearchShortcut)
      document.removeEventListener('keydown', handleOnSearchShortcut)
    }
  }, [handleOnSearchShortcut])

  return (
    <div className={moduleStyles.autocompleteSearchBox}>
      <SearchBox
        ref={searchBoxRef}
        defaultValue={defaultValue}
        placeholder={placeholder}
        onChange={onChange}
      />
      {isInputFocused && options.length > 0 ? (
        <div className={classNames([moduleStyles.autocompleteSuggestions, styles.my2])}>
          <ListGroup>
            {options.map((option, idx) => (
              <ListGroupItem key={idx} colorStyle={focusIndex === idx ? focusStyle : undefined}>
                {renderOption(option, idx, focusIndex)}
              </ListGroupItem>
            ))}
            {renderFooter && (
              <ListGroupItem key="footer" colorStyle="secondary">
                {renderFooter(options)}
              </ListGroupItem>
            )}
          </ListGroup>
        </div>
      ) : null}
    </div>
  )
}
