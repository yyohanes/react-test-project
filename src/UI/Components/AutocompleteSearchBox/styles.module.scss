.autocomplete-search-box {
  position: relative;

  &__suggestions {
    position: absolute;
    width: 100%;
    z-index: 1;
  }

  &__results {
    max-height: 50vh;
    overflow: auto;
  }
}
