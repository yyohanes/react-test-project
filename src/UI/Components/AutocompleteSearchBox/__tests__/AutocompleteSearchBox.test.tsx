import React from 'react'
import {
  render,
  act,
  fireEvent,
  getByText,
} from '@testing-library/react'

import AutocompleteSearchBox from '..'

describe('AutocompleteSearchBox', () => {
  const testOptions = [
    {
      title: 'Test 1',
    },
    {
      title: 'Test 2',
    },
  ]

  test('Should not render suggestions if input is not focused', () => {
    const component = render(
      <AutocompleteSearchBox
        placeholder="Search"
        defaultValue="Test"
        options={testOptions}
        renderOption={option => (<p>{option.title}</p>)}
      />,
    )

    expect(component.container.querySelectorAll('.option-item').length).toBe(0)
  })

  test('Should render suggestions if options > 0 & input is focused', () => {
    const component = render(
      <AutocompleteSearchBox
        placeholder="Search"
        options={testOptions}
        renderOption={option => (<p className="option-item">{option.title}</p>)}
      />,
    )
    act(() => {
      component.getByPlaceholderText('Search').focus()
    })

    expect(component.container.querySelectorAll('.option-item').length).toBe(2)
    expect(component.container).toMatchSnapshot()
  })

  test('Should select correct auto-suggestion when navigating with ArrowUp or ArrowDown', () => {
    const component = render(
      <AutocompleteSearchBox
        placeholder="Search"
        options={testOptions}
        focusStyle="primary"
        renderOption={(option, idx, focusIndex) => (<p className={`option-item ${idx === focusIndex ? 'focused' : ''}`}>{option.title}</p>)}
      />,
    )
    const input = component.getByPlaceholderText('Search')
    act(() => {
      input.focus()
      fireEvent.keyDown(input, { key: 'ArrowDown', keyCode: 40 })
    })

    const expectedActiveElement = getByText(component.container, 'Test 1')
    const expectedActiveElement2 = getByText(component.container, 'Test 2')

    expect(expectedActiveElement.className).toBe('option-item focused')

    act(() => {
      fireEvent.keyDown(input, { key: 'ArrowDown', keyCode: 40 })
    })

    expect(expectedActiveElement.className).toBe('option-item ')
    expect(expectedActiveElement2.className).toBe('option-item focused')

    act(() => {
      fireEvent.keyDown(input, { key: 'ArrowUp', keyCode: 38 })
    })
    expect(expectedActiveElement2.className).toBe('option-item ')
    expect(expectedActiveElement.className).toBe('option-item focused')
  })
})
