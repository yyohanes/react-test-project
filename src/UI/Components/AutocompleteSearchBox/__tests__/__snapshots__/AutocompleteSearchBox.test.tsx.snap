// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteSearchBox Should render suggestions if options > 0 & input is focused 1`] = `
<div>
  <div
    class="autocompleteSearchBox"
  >
    <input
      class=""
      placeholder="Search"
      type="text"
      value=""
    />
    <div
      class="autocompleteSuggestions"
    >
      <div>
        <div
          class=""
        >
          <p
            class="option-item"
          >
            Test 1
          </p>
        </div>
        <div
          class=""
        >
          <p
            class="option-item"
          >
            Test 2
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
